manpages/attbckshare.1
.\"Text automatically generated by txt2man
.TH attbckshare 1 "12 May 2014" "" "Backup Scripts Manual"
.SH NAME
\fBattbckshare \fP- attach network share for backups
.SH SYNOPSIS
.nf
.fam C
\fBattbckshare\fP [\fIOPTION\fP]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
Attaches a pre-determined network samba share as a target location for backup
files.
.SH OPTIONS
\fB-h\fP prints usage information
.PP
\fB-v\fP verbose output
.PP
\fB-V\fP prints version information
.SH EXIT STATUS
0
if OK,
.PP
1
if minor problems,
.PP
64
error in command line arguments.
.PP
65
failed to mount share due to mount error.
.PP
66
Backup share already mounted.
.SH AUTHOR
Written by Mark Grant.
.SH REPORTING BUGS
Report bugs to <m.grant.prg@btinternet.com>
.SH COPYRIGHT
Copyright 2010 Mark Grant.
.br
License GPLv3+: GNU GPL version 3 or later
.br
<http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
\fBchkbckshare\fP(1) \fBdetbckshare\fP(1) \fBmount\fP(8) (and \fBmount_smbfs\fP(8) on FreeBSD) \fBumount\fP(8)


NEWS
Release 1.1.1 2014-01-20
Changed stdout & stderr message output to use a function directing to one or
other based on a status. Changed version option to -V. Added usage of
standard error routine. Introduced function to repeat commands to overcome
latency issues on NAS mount from either mount_smbfs or slow hardware.

Release 1.0.27 2013-10-30
Removed '-' prepending operations in the tar command. Added --exclude-caches
to the tar command. Also changed backup command in bckfulladhoc to use tar or
gtar depending on OS.

Release 1.0.26 2013-09-14
Logic error introduced in bckfullweekly v1.0.19 corrected.

Release 1.0.25 2013-09-14
Improved coding around sequence calculation.

Release 1.0.24 2013-09-14
Changed sequence calculation from DOM / 7 + 1 to testing day
against 7, 14, 21 and 28 as original calc did not work correctly when
day was exactly divisible by 7.

Release 1.0.23 2013-08-22
Debian best practice documentation requires that programs placed in
the PATH should not have extensions as the programs may be rewritten
in a different language. This would then require a name change, or,
if left as-is would cause confusion. So all names changed to remove .sh
extension except for setup.sh which is not on the PATH.

Release 1.0.22 2013-04-01
Changed all scripts to use new backups sub-directory of sysconfdir.Added
include and exclude from files.

Release 1.0.21 2013-03-02
Created new script to support backups from a stored list of files to backup.

Release 1.0.20 2013-02-26
Changed internals of processing command line options.

Release 1.0.19 2013-02-06
Added mailing of backup file date hierachy after backup.

Release 1.0.18 2012-12-20
Added Host name and backup file indicator to email message subject line.

Release 1.0.17 2012-11-17
Changed logic to give responsibility to daily backup to delete the previous
version of that days backup so we now have a full week of full + incremental
backups.

Release 1.0.16 2012-11-06
Reorganised manual pages directory structure in source.

Release 1.0.15 2012-11-06
Updated Author contact information. Moved installation from bindir to sbindir.

Release 1.0.14 2012-11-06
Reverted to use the .sh file extension. Added exclusion to tar command for /run
and /var/run following inclusion of /run in Linux.

Release 1.0.13 2012-01-10
Removed the .sh extension from the command name. Add .gvfs file exclusion to
support Gnome desktops and Ubuntu.

Release 1.0.12 2010-12-16
Allowed mailing of disk usage stats from all OS's

Release 1.0.11 2010-12-14
Modification to an invalid mail switch.

Release 1.0.10 2010-11-28
Modifications to produce and use configuration files for accessing the NAS.

Release 1.0.9 2010-11-23
Daily differential backups for Linux have been restored following a tar bug
fix.

Release 1.0.8 2010-11-21
Re-vamped all manual pages.

Release 1.0.7 2010-11-20
Added manual pages for all commands.

Release 1.0.6 2010-11-20
The daily and weekly scripts no longer call shutdown on cmpletion. Shutdown
must be invoked manually.

Release 1.0.5 2010-11-18
First release of the backup utility scripts.


scripts/attbckshare.sh
#! /usr/bin/env bash
##########################################################################
##									##
##	attbckshare.sh is automatically generated,			##
##		please do not modify!					##
##									##
##########################################################################

##########################################################################
##									##
## Script ID: attbckshare.sh						##
## Author: Mark Grant							##
##									##
## Purpose:								##
## To mount the backup share on the NAS server. E.g.			##
## 	\\Ambrosia\charybdisbck						##
## 	on								##
## 	/mnt/charybdisbck						##
##									##
## Syntax:	attbckshare.sh [-h || --help || -v || --version]	##
##									##
## Exit Codes:	0 & 64 - 113 as per C/C++ standard			##
##		0 - success						##
##		64 - Invalid arguments					##
##		65 - Failed mounting backup NAS server			##
##		66 - Backcup share already mounted			##
##		67 - trap received					##
##									##
## Further Info:							##
## This script mounts a NAS share as a target for the backup scripts.	##
##									##
## In order to make the package portable all the necessary parameters	##
## are stored in a $PREFIX/etc/backups.conf file.			##
## On FreeBSD mounting a NAS share uses the mount_smbfs command rather	##
## than the mount.cifs command on Linux. This difference means that on	##
## FreeBSD we also utilise the ~/.nsmbrc file.				##
## The format of the backups.conf file is below:			##
##	# NAS server name						##
##	bckupsys=MyServer						##
##									##
##	# NAS directory for backups (also used for mount point eg	##
##	#				/mnt/$bckupdir)			##
##	bckupdir=mybackupdirectory					##
##									##
##	# NAS backup user						##
##	bckupusr=mybackupuser						##
##									##
##	# NAS backup user password					##
##	bckuppwd=mybackuppassword					##
##									##
##	# Backup workgroup						##
##	bckupwg=MyWorkgroup						##
##									##
##	# Notify user							##
##	notifyuser=root							##
##									##
## The format of the ~/.nsmbrc file is below:-				##
##	# First define a workgroup.					##
##	[default]							##
##	workgroup=MyWorkgroup						##
##									##
##	# Then define a server.						##
##	[MYSERVER]							##
##	addr=MyServer							##
##									##
##	# Then define a server / user pair.				##
##	[MYSERVER:MYBACKUPUSER]						##
##	# Use persistent password cache for user 'mybackupuser'		##
##			on system 'MyServer'.				##
##	password=mybackuppassword					##
##									##
##########################################################################

##########################################################################
##									##
## Changelog								##
##									##
## Date		Author	Version	Description				##
##									##
## 09/04/2010	MG	1.0.1	Created for Linux.			##
## 26/08/2010	MG	1.0.2	Major revision completed supporting	##
##				FreeBSD as well as Linux. Also put all	##
##				relevant parameters in variables at	##
##				beginning of script to enhance		##
##				portability. (e.g. System, backup user	##
##				etc.).					##
## 18/11/2010	MG	1.0.3	Changed to emit help and version on	##
##				input of correct flag as argument. Also	##
##				stored version in string in Init section##
## 28/11/2010	MG	1.0.4	Changed script to read parameters from	##
##				etclocation/backups.conf and, when	##
##				necessary (FreeBSD), the ~/.nsmbrc file.##
## 10/01/2012	MG	1.0.5	Removed the .sh extension from the	##
##				command name.				##
## 05/11/2012	MG	1.0.6	Reverted to the .sh file extension.	##
## 26/02/2013	MG	1.0.7	Changed command line option processing	##
##				to use getopts.				##
## 01/04/2013	MG	1.0.8	Moved config files to new backups	##
##				directory under sysconfdir. Added	##
##				exclude files for system, weekly, daily	##
##				and data.				##
##									##
##########################################################################

####################
## Init variables ##
####################
script_exit_code=0
osname=$(uname -s)			# Get system name for OS differentiation
version="1.0.8"				# set version variable
etclocation=/usr/local/etc/backups	# Path to etc directory

###############
## Functions ##
###############
# Standard function to log $1 to stderr
mess_log()
{
echo $1 1>&2
}

# Standard function to message stderr, any cleanup and return exit code
script_exit()
{
mess_log "$1" 
exit $script_exit_code
}

# Standard trap exit function
trap_exit()
{
script_exit_code=67
script_exit "Script terminating due to trap received. Code: "$script_exit_code
}

# Setup trap
trap trap_exit SIGHUP SIGINT SIGTERM

##########
## Main ##
##########
# Process command line arguments with getopts.
while getopts :hv arg
do
	case $arg in
		h)	echo "Usage is $0 [options]"
			echo "	-h displays usage information"
			echo "	OR"
			echo "	-v displays version information"
			;;
		v)	echo "$0 version "$version
			;;
		\?)	echo "Invalid argument -$OPTARG." >&2
			exit 64
			;;
	esac
done

# If help or version requested then exit now.
if [ $# -gt 0 ]
	then
		exit 0
fi

# Read parameters from $etclocation/backups.conf
IFS="="

exec 3<$etclocation/backups.conf
while read -u3 -ra input
do
	case ${input[0]} in
	server)
		bckupsys=${input[1]}
		;;
	dir)
		bckupdir=${input[1]}
		;;
	user)
		bckupusr=${input[1]}
		;;
	password)
		bckuppwd=${input[1]}
		;;
	esac
done
exec 3<&-

# If FreeBSD we also need the workgroup which can be obtained from
# the ~/.nsmbrc file

if [ $osname = "FreeBSD" ]; then
	exec 3<~/.nsmbrc
	while read -u3 -ra input
	do
		case ${input[0]} in
		workgroup)
			bckupwg=${input[1]}
			;;
		esac
	done
	exec 3<&-
fi

# Check to see if the NAS backup server is mounted, if not, mount
if [ "$(mount | grep "/mnt/$bckupdir")" == "" ] 
	then
	case $osname in
	FreeBSD)
		mount_smbfs -I $bckupsys -N -U $bckupusr -W $bckupwg \
        		//$bckupusr@$bckupsys/$bckupdir /mnt/$bckupdir
		status=$?
	;;
	Linux)
	mount -t cifs -o user=$bckupusr,password=$bckuppwd \
        	//$bckupsys/$bckupdir /mnt/$bckupdir
	status=$?
	;;
	esac
	if [ $status != "0" ]
		then
		script_exit_code=65
		script_exit "Failed to mount backup NAS server. Mount error: "$status" Script exit code: "$script_exit_code
	fi
	else
	script_exit_code=66
	script_exit "Backup share already mounted. Script exit code: "$script_exit_code
fi
exit 0
